src/libsyntax/codemap.rs

```diff
@@ -1020,6 +1157,123 @@ impl CodeMap {
+            let cur_lo = self.lookup_char_pos(sp.lo);
+            if prev_hi.line == cur_lo.line {
+                push_trailing(&mut buf, prev_line, &prev_hi, Some(&cur_lo));
+            } else {
+                push_trailing(&mut buf, prev_line, &prev_hi, None);
+                // push lines between the previous and current span (if any)
+                for idx in prev_hi.line..(cur_lo.line - 1) {
+                    if let Some(line) = fm.get_line(idx) {
+                        buf.push_str(line);
+                        buf.push('\n');
+                    }
+                }
+                if let Some(cur_line) = fm.get_line(cur_lo.line - 1) {
+                    buf.push_str(&cur_line[.. cur_lo.col.to_usize()]);
+                }
+            }
+            buf.push_str(substitute);
+            prev_hi = self.lookup_char_pos(sp.hi);
+            prev_line = fm.get_line(prev_hi.line - 1);
+        }
+        push_trailing(&mut buf, prev_line, &prev_hi, None);
+        // remove trailing newline
+        buf.pop();
+        buf
+    }
+
+    pub fn custom_group_spans<F>(&self, mut spans: Vec<Span>, push: F) -> Vec<MultiSpan>
+    where F: Fn(&mut MultiSpan, Span) {
+        spans.sort_by(|a, b| a.lo.cmp(&b.lo));
+        let mut groups = Vec::<MultiSpan>::new();
+        let mut overflowing = vec![];
+        let mut prev_expn = ExpnId(!2u32);
+        let mut prev_file = !0usize;
+        let mut prev_line = !0usize;
+        let mut err_size = 0;
+
+        for sp in spans {
+            let line = self.lookup_char_pos(sp.lo).line;
+            let line_hi = self.lookup_char_pos(sp.hi).line;
+            if line != line_hi {
+                overflowing.push(sp.into());
+                continue
+            }
+            let file = self.lookup_filemap_idx(sp.lo);
+
+            if err_size < MAX_HIGHLIGHT_LINES && sp.expn_id == prev_expn && file == prev_file {
+                // `push` takes care of sorting, trimming, and merging
+                push(&mut groups.last_mut().unwrap(), sp);
+                if line != prev_line {
+                    err_size += 1;
+                }
+            } else {
+                groups.push(sp.into());
+                err_size = 1;
+            }
+            prev_expn = sp.expn_id;
+            prev_file = file;
+            prev_line = line;
+        }
+        groups.extend(overflowing);
+        groups
+    }
+
+    /// Sort, group by lines, and merge overlapping spans.
+    /// Spans overflowing a line are put into their own group.
+    pub fn group_spans(&self, spans: Vec<Span>) -> Vec<MultiSpan> {
+        self.custom_group_spans(spans, |msp, sp| msp.push_merge(sp))
+    }
+
+    /// Like `group_spans`, but trims overlapping spans instead of
+    /// merging them (for use with `end_highlight_lines`)
+    pub fn end_group_spans(&self, spans: Vec<Span>) -> Vec<MultiSpan> {
+        self.custom_group_spans(spans, |msp, sp| msp.end_push(sp))
+    }
+
```

Comment:
I think all these new methods would be better in with error emitters, rather than in the codemap. In particular, how we treat multispans might be different between different emitters.